runtime/warp.cu

```diff
@@ -88,4 +88,58 @@ __device__ void warpReduceTIDX(
   }
 }
 
+template <int BDIMX, int BDIMY, bool Aligned, typename T, typename Func>
```

Comment:
Why do we need to have this template function with a slight different implementation for inter-warp reduction?

Reply:
The existing version `warpReduceTIDX` is doing reduction in X dim, Y dim is used for iteration. It requires `bdimx % 32 ==0`
This new version `warpReduceTIDXY` is doing reduction in X & Y dims, it requires `(bdimx * bdimy) % 32 ==0`. So they are different. Also `bdimx` and `bdimy` are static values, then we can use template vars.

Reply:
ah, my bad. I didn't realized that in warpReduceTIDX, we are still using threadIdx.y. Earlier I was suggesting that these two should be merged together, but that doesn't look like worth it.